File: playground/mt/sample/mt.rb
W = 32
N = 634
M = 397
R = 31
U = 11
S = 7
T = 15
L = 18
A = 0x9908B0DF
B = 0x9D2C5680
C = 0xEFC60000

WHOLE_MASK = ("1" * W).to_i(2)
UPPER_MASK = ("1" * (W-R) + "0" * R).to_i(2)
LOWER_MASK = ("0" * (W-R) + "1" * R).to_i(2)

class MT
  def initialize(seed)
    @i = 0
    @x = [seed & WHOLE_MASK]

    1.upto(N-1) do |i|
      @x[i] = (1812433253 * (@x[i-1] ^ (@x[i-1] >> 30)) + i) & WHOLE_MASK
    end
    p @x
  end

  def next
    z = @x[@i] & UPPER_MASK | @x[(@i + 1) % N] & LOWER_MASK
    
    @x[@i] = @x[(@i + M) % N] ^ (z >> 1) ^ (z & 1 == 0 ? 0 : A)
    
    y = @x[@i]
    y = y ^ (y >> U)
    y = y ^ ((y << S) & B)
    y = y ^ ((y << T) & C)
    y = y ^ (y >> L)
  
    @i = (@i + 1) % N
    return y
  end
end

mt = MT.new(20150919)
2048.times do |i|
  print i, ": ", mt.next, "\n"
end

File: playground/mt/sample/ocaml/mt.ml
(* MT19937 *)

let w = 32;;
let n = 624;;
let m = 397;;
let r = 31;;
let u = 11;;
let s = 7;;
let t = 15;;
let l = 18;;
let a = 0x9908B0DF;;
let b = 0x9D2C5680;;
let c = 0xEFC60000;;

let whole_mask = int_of_float (2.0 ** (float_of_int (w))) - 1;;
let upper_mask = int_of_float (2.0 ** (float_of_int (w-1)));;
let lower_mask = upper_mask - 1;;

type mtRand = {i: int; x: int list};;

let initialize seed =
  let rec generate acc index =
    if index = n then acc
    else generate ((1812433253 * ((List.hd acc) lxor ((List.hd acc) lsr 30)) + index) land whole_mask::acc) (index+1)
  in let l = generate [seed land whole_mask] 1
  in {i = 0; x = List.rev l};;

let rec update lst n x =
  if n = 0 then x :: (List.tl lst)
  else (List.hd lst) :: update (List.tl lst) (n-1) x;;

let next rand =
  let z = ((List.nth rand.x rand.i) land upper_mask) lor ((List.nth rand.x ((rand.i + 1) mod n)) land lower_mask) in
  let xi = (List.nth rand.x ((rand.i + m) mod n)) lxor (z lsr 1) lxor (if z land 1 = 0 then 0 else a) in
  let y0 = xi in
  let y1 = y0 lxor (y0 lsr u) in
  let y2 = y1 lxor ((y1 lsl s) land b) in
  let y3 = y2 lxor ((y2 lsl t) land c) in
  let y4 = y3 lxor (y3 lsr l) in
  (y4, {i = (rand.i + 1) mod n; x = update rand.x rand.i xi});;


let init = initialize 20150919;;

let printRandomSeq seed n =
  let rec iter index rand =
    if index = n then ()
    else
      let (nextInt, nextRand) = next rand in
      begin
        Printf.printf "%d:%d\n" index nextInt;
        iter (index+1) nextRand
      end
  in iter 0 (initialize seed);;
  
printRandomSeq 20190820 2048;;
